### DISCUSSION TITLE
How to use the output of one tool as input for another tool

### DISCUSSION THREAD

**Discussion by eliasfroehner** [patterns-ai-core/langchainrb#733] [Q&A] [answered]
I want to know how I can use the result of one tool as input for another tool. 

Below is my current source code:

```ruby
llm = Langchain::LLM::OpenAI.new(
  api_key: "na",
  llm_options: {uri_base: "http://llm.ollama.demo:11434/", log_errors: true},
  default_options: {chat_completion_model_name: "llama3.1"}
)
thread = Langchain::Thread.new
assistant = Langchain::Assistant.new(
  llm: llm,
  thread: thread,
  instructions: "You are a smart general assistant able to use tools if needed.",
  tools: [Langchain::Tool::FileSystem.new, Langchain::Tool::Wikipedia.new]
)
user_message = "Write a summary about Steve Jobs and write it to summary.md"
assistant.add_message_and_run(content: user_message, auto_tool_execution: true)
```

The LLM returns two tool calls. However, `Langchain::Tool::FileSystem#write_to_file` is already populated with content from the LLM's general knowledge, not from the summary provided by Wikipedia.

How can I ensure that the output from the Wikipedia tool is used as input for the FileSystem tool?

**Comment by andreibondarev** [comment]
@eliasfroehner Is using a local LLM a deal breaker for you? I just tried it out with OpenAI and it worked.

**Reply by eliasfroehner** [reply]
@andreibondarev Yes, unfortunately, it is, because company policies only allow the use of local LLMs. However, it also works immediately for me with OpenAI. I tried it with mistral-nemo and the llama 3.1 models from Groq—always the same. While debugging, I noticed that GPT-4o Mini executes tool calls sequentially, whereas Llama 3.1 returns both tool calls at once. I wonder if Ollama could be the problem...

**Reply by andreibondarev** [reply]
@eliasfroehner Let me try something and report back.

**Reply by eliasfroehner** [reply]
@andreibondarev Thank you very much.

**Comment by andreibondarev** [comment, marked as answer]
I strung the 2 assistants (research assistant + file system assistant) together and it worked.
```ruby
# This worked better than the OpenAI class using the Ollama URL
llm = Langchain::LLM::Ollama.new # default is chat_completion_model_name: "llama3.1"

# Instantiate researcher_assistant
researcher_assistant = Langchain::Assistant.new(
  llm: llm,
  # No need to pass a Langchain::Thread.new, it gets auto-instantiated.
  instructions: "You are a smart general researcher that uses Wikipedia to write summaries on various topics.",
  tools: [Langchain::Tool::Wikipedia.new]
)

# Instantiate fs_assistant
fs_assistant = Langchain::Assistant.new(
  llm: llm,
  instructions: "You are a file system assistant.",
  tools: [Langchain::Tool::FileSystem.new]
)

# Invoke the researcher_assistant
messages = researcher_assistant.add_message_and_run(
  content: "Write a summary about Steve Jobs",
  auto_tool_execution: true
)

# If successful, last message should be the summary
summary = messages.last.content if researcher_assistant.state == :completed

# Feed this into the fs_assistant
messages = fs_assistant.add_message_and_run(
  content: "Write this to summary.md: #{summary}",
  auto_tool_execution: true
)

# Inspect the last message
messages.last
```

Would you please give this a spin? I'm still trying to figure out the best approach to build these workflows togerher.

**Reply by eliasfroehner** [reply]
Thank you! I think it is the only solution with open source LLM’s at the moment.

**Reply by andreibondarev** [reply]
GoogleGemini, OpenAI and Anthropic support a `tool_choice:` parameter that forces an LLM to use specific tool. Ollama doesn't (yet).

Once it's supported -- I was thinking that you could dynamically set tool_choice on the assistant. In theory you'd be able to do something like:
```ruby
assistant = Langchain::Assistant.new(
  llm: llm,
  instructions: "You are a smart general assistant able to use tools if needed.",
  tools: [Langchain::Tool::FileSystem.new, Langchain::Tool::Wikipedia.new],
  tool_choice: "wikipedia__execute"
)

assistant.add_message_and_run(
  content: "Write a summary about Steve Jobs",
  auto_tool_execution: true
)

# Force a different tool to be used
assistant.tool_choice = "file_system__write_to_file"

assistant.add_message_and_run(
  content: "Write it to summary.md",
  auto_tool_execution: true
)
```

**Reply by eliasfroehner** [reply]
OK, that’s a good idea.

It would be interesting to see how Ollama behaves in its OpenAI compatibility layer regarding this feature.

**Reply by andreibondarev** [reply]
@eliasfroehner I added the ticket https://github.com/patterns-ai-core/langchainrb/issues/738 to address this.

**Reply by eliasfroehner** [reply]
Great! Thank you
